hs-abci-examples/nameservice/src/Nameservice/Modules/Nameservice/Types.hs

```diff
@@ -1,42 +1,52 @@
-import           Data.String.Conversions        (cs)
-import           Data.Text                      (Text)
-import           GHC.Generics                   (Generic)
-import           Nameservice.Aeson              (defaultNameserviceOptions)
-import           Nameservice.Modules.Token      (Address (..), Amount (..))
-import qualified Proto.Nameservice.Whois        as W
-import qualified Proto.Nameservice.Whois_Fields as W
-import           Tendermint.SDK.Auth            (Address, addressFromBytes,
-                                                 addressToBytes)
-import           Tendermint.SDK.Codec           (HasCodec (..))
-import           Tendermint.SDK.Errors          (AppError (..), IsAppError (..))
-import           Tendermint.SDK.Events          (FromEvent (..), ToEvent (..))
-import qualified Tendermint.SDK.Router          as R
-import qualified Tendermint.SDK.Store           as Store
+import           Control.Lens              (iso)
+import           Data.Aeson                as A
+import qualified Data.Binary               as Binary
+import           Data.String.Conversions   (cs)
+import           Data.Text                 (Text)
+import qualified Data.Text.Lazy            as TL
+import           GHC.Generics              (Generic)
+import           Nameservice.Aeson         (defaultNameserviceOptions)
+import           Nameservice.Modules.Token (Amount (..))
+import           Proto3.Suite              (HasDefault, Message, MessageField,
+                                            Named, Primitive (..))
+import qualified Proto3.Suite.DotProto     as DotProto
+import qualified Proto3.Wire.Decode        as Decode
+import qualified Proto3.Wire.Encode        as Encode
+import           Tendermint.SDK.Auth       (Address)
+import           Tendermint.SDK.Codec      (HasCodec (..))
+import           Tendermint.SDK.Errors     (AppError (..), IsAppError (..))
+import           Tendermint.SDK.Events     (FromEvent (..), ToEvent (..))
+import qualified Tendermint.SDK.Router     as R
+import qualified Tendermint.SDK.Store      as Store
 
 --------------------------------------------------------------------------------
 
 type NameserviceModule = "nameservice"
 
 --------------------------------------------------------------------------------
 
-newtype Name = Name String deriving (Eq, Show, Binary.Binary, A.ToJSON, A.FromJSON)
+newtype Name = Name Text deriving (Eq, Show, Generic, Binary.Binary, A.ToJSON, A.FromJSON)
+instance Primitive Name where
+  encodePrimitive n (Name txt) = Encode.text n . TL.fromStrict $ txt
+  decodePrimitive = Name . TL.toStrict <$> Decode.text
+  primType _ = DotProto.String
+
+instance HasDefault Name
+
+instance MessageField Name
 
 data Whois = Whois
-  { whoisValue :: String
+  { whoisValue :: Text
   , whoisOwner :: Address
   , whoisPrice :: Amount
   } deriving (Eq, Show, Generic)
 
+instance Message Whois
+
+instance Named Whois
+
 instance Binary.Binary Whois
 
 instance HasCodec Whois where
```

Comment:
You should use the binary codecs from protobuf here and remove `Binary` definitions